[LeeEnEn] iP

LeeEnEn opened the pull request:
## Duke Bot
> "All problems are illusions of the mind" - Eckhart Tolle

Duke Bot frees your mind of having to remember things you need to do. It is,
- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2122S2/ip)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :smiley:

And it is **FREE!**

Features
- [x] Managing tasks
- [ ] Managing deadline (soon)
- [ ] Reminders (soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

``` ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Level 1. Greet, Echo, Exit.

Commit:
Level 2. Add, List.

Commit:
Level 3. Mark as Done.

Commit:
Level 4. ToDos, Events, Deadlines.

Commit:
A-TextUiTesting

Commit:
A-Exceptions.

Commit:
A-Collections.

Commit:
A-Enums.

Commit:
Update missing Enum

Commit:
Level 7. Save

Commit:
Level 8. Dates and Times

Commit:
Revert "Level 8. Dates and Times"

This reverts commit 8f16d0105d84daccf042a267fd29bbe77bb8f3a8.

Commit:
Merge branch 'branch-Level-7'

Commit:
Level 8. Dates and times

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java

Commit:
Update switch statements to comply with coding standard

Commit:
Update DukeFile to output LocalDate instead of String

Commit:
A-MoreOOP: Use More OOP

Rename DukeFile to Storage

Commit:
Add packages to files

Commit:
Fix storage package to have small letters

Commit:
A-JUnit: Add JUnit Tests

Commit:
A-JavaDoc: JavaDoc

Commit:
A-CodingStandard: Follow the Coding Standard

Commit:
Level 9. Find

Users can now find a task by searching for a keyword.
Find command is case-sensitive.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

channne commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,194 @@
+        Type type = getEnumType(input[0]);
+        TaskList taskList = new TaskList();
+
+        try {
```
I really like how readable this block is, as well as how you printed different types of error messages for different types of errors :smiley:

LeeEnEn replied:
Thank you! 😄

channne commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,194 @@
+        } catch (DukeException e) {
+            if (isCommandRecognized(userInput[0])) {
+                ui.echo("☹ OOPS!!! The description of a " + userInput[0] + " cannot be empty.");
+            } else if (e.getMessage().equals("Error in task") || e.getMessage().equals("Error")) {
```
Perhaps it would be better to create different classes of exceptions instead of recognizing the different types of exceptions by their error messages?

LeeEnEn replied:
That makes sense, I will do the necessary changes!

channne commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,88 @@
+ * Handles the creation of tasks.
+ */
+public class Task {
+    private boolean marked;
```
Consider renaming the booleans in this class?

LeeEnEn replied:
Good catch! Definitely missed this during my checks!

channne commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,194 @@
+     * @param input User command.
+     * @return True if Duke understands the command, false otherwise.
+     */
+    private boolean isCommandRecognized(String input) {
```
Perhaps this method could be better named, since it detects four specific types of commands rather than all commands

LeeEnEn replied:
You are right, thank you for pointing that out! 😄

channne commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,94 @@
+            }
+            scanner.close();
+        } catch (FileNotFoundException e) {
+            System.out.println("OOPS!!! Duke.Main.Duke is unable to locate your file!");
```
Perhaps it would be better to print error messages through the UI as well?

LeeEnEn replied:
This makes sense too! Thank you for the feedback! 👍

channne reviewed (commented):
Overall, I think your code is very readable and I could hardly find any coding standard violations :+1: One thing I noticed however is that many classes communicate with the UI! Perhaps it might be beneficial to restructure such that fewer classes communicate with it :smile:

brandonrhan commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,194 @@
+     * @return A boolean.
+     */
+    public boolean isExit() {
+        return this.isExit;
```
You never use "this." in previous code, maybe you could delete this one as well

brandonrhan commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class Deadline extends Task {
+
+    private LocalDate by;
```
Maybe you should rename it using a more detailed name

brandonrhan commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString()
+                + " (by: " + by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
The pattern could be saved as a static variable to avoid duplicate code

brandonrhan commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class Event extends Task {
+
+    private LocalDate at;
```
at is also not very clear, I think could change to another name

brandonrhan commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,88 @@
+     * @param mark A boolean to mark or unmark a task.
+     * @param show A boolean to print Duke's reply to the user.
+     */
+    public void markTask(boolean mark, boolean show) {
```
show should also be renamed

brandonrhan reviewed (commented):
I really like your code!! Your naming of variables is so good and your code is extremely clear to read. I have learned a lot from you!!

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Level-10: GUI

Change from text based UI to GUI

Commit:
Update code to comply with Java Coding Standard

Commit:
Update UI

Replace current stock DaDuke and DaUser to custom pictures, default text font to Verdana and text size to 15 for better readability.

Change maximum height of VBox in MainWindow.fxml to infinity, preventing text from being cut off.

Commit:
A-Assertions: Use Assertions

Commit:
A-CodeQuality: Improve Code Quality

LeeEnEn committed:
Merge pull request #2

Add assertions for necessary classes #1

LeeEnEn committed:
Merge branch 'master' into branch-A-CodeQuality

LeeEnEn committed:
Merge pull request #3 from LeeEnEn/branch-A-CodeQuality

Improve code quality #2

Commit:
A-CI: Set up CI

Commit:
C-DetectDuplicates

Commit:
Add missing checkstyle

Update files to comply with checkstyle after adding required checkstyle files.

Commit:
A-BetterGui: Better GUI

Improve GUI looks.

Commit:
Update Ui

Increase height and width of application, reduce picture size of Duke and user and also added application icon.

Commit:
Update missing response.

Commit:
Update missing response.

Commit:
A-UserGuide: User Guide

Commit:
Update Ui

Removed extra spacing in list items after calling find function.

Allowing deadline and event type tasks to take in '/' in their descriptions.

Commit:
Revert accidental changes

Commit:
Change mainClassName to correct value

This allows the execution of the jar file.
